src/data/list/alist.lean

```diff
@@ -241,6 +245,50 @@ ext $ by simp only [alist.insert_entries, list.kerase_cons_eq, and_self, alist.s
 theorem to_alist_cons (a : α) (b : β a) (xs : list (sigma β)) :
   list.to_alist (⟨a,b⟩ :: xs) = insert a b xs.to_alist := rfl
 
+theorem mk_cons_eq_insert (c : sigma β) (l : list (sigma β)) (h : (c :: l).nodupkeys) :
+  (⟨c :: l, h⟩ : alist β) = insert c.1 c.2 ⟨l, nodupkeys_of_nodupkeys_cons h⟩ :=
+by simpa [insert] using (kerase_of_not_mem_keys $ not_mem_keys_of_nodupkeys_cons h).symm
+
+/-- Recursion on an `alist`, using `insert`. Use as `induction l using alist.insert_rec`. -/
+@[elab_as_eliminator] def insert_rec {C : alist β → Sort*} (H0 : C ∅)
+  (IH : Π (a : α) (b : β a) (l : alist β) (h : a ∉ l), C l → C (l.insert a b)) : Π l : alist β, C l
+| ⟨[], _⟩     := H0
+| ⟨c :: l, h⟩ := begin
+  rw mk_cons_eq_insert,
```

Comment:
It feels very suspicious to me that this is defining data in tactic mode, in particular starting with a rewrite. Is there no way to give this data in term mode, without using a tactic like rewrite first?

Reply:
In theory this shouldn't matter, since the theorems `insert_rec_empty` and `insert_rec_insert` tell us that this data behaves exactly as expected. We could avoid the `rw` and write down the casts explicitly, but I can't see how that's any better. 

The problem here is that cons on the underlying list of an `alist` being equal to `insert` is nowhere near def-eq. I can't see any nicer way around this.